pandas/core/indexes/base.py

```diff
@@ -301,11 +301,17 @@ def __new__(cls, data=None, dtype=None, copy=False, name=None,
               (dtype is not None and is_datetime64_any_dtype(dtype)) or
                 'tz' in kwargs):
             from pandas import DatetimeIndex
-            result = DatetimeIndex(data, copy=copy, name=name,
-                                   dtype=dtype, **kwargs)
+
             if dtype is not None and is_dtype_equal(_o_dtype, dtype):
-                return Index(result.to_pydatetime(), dtype=_o_dtype)
+                # GH#23524 passing `dtype=object` to DatetimeIndex is invalid,
+                #  will raise in the where `data` is already tz-aware.  So
+                #  we leave it out of this step and cast to object-dtype after
+                #  the DatetimeIndex construction.
+                result = DatetimeIndex(data, copy=copy, name=name, **kwargs)
+                return Index(list(result), dtype=_o_dtype)
```

Comment:
I would say you are fixing the sympton on not the problem here. if passing object dtype, then NO conversion should be done like this. having to pass a list is very odd (the list-ifying should be done *inside* the constructor)

Reply:
When you say "_inside_ the constructor", which constructor are you referring to?  This _is_ inside the `Index.__new__` constructor.  Would you prefer that the `return Index(list(result), dtype=_o_dtype)` line were changed to `return result.astype(object)`?

Really I'm just not clear on what you're looking for here.